server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java

```diff
@@ -191,30 +192,39 @@ private void updateRecoveryCounts(final ShardRouting routing, final boolean incr
             return;
         }
 
-        Recoveries.getOrAdd(recoveriesPerNode, routing.currentNodeId()).addIncoming(howMany);
+        Recoveries.getOrAdd(getReplicaRecoveries(routing), routing.currentNodeId()).addIncoming(howMany);
 
         if (routing.recoverySource().getType() == RecoverySource.Type.PEER) {
             // add/remove corresponding outgoing recovery on node with primary shard
             if (primary == null) {
                 throw new IllegalStateException("shard is peer recovering but primary is unassigned");
             }
-            Recoveries.getOrAdd(recoveriesPerNode, primary.currentNodeId()).addOutgoing(howMany);
+
+            Recoveries.getOrAdd(getReplicaRecoveries(routing), primary.currentNodeId()).addOutgoing(howMany);
 
             if (increment == false && routing.primary() && routing.relocatingNodeId() != null) {
                 // primary is done relocating, move non-primary recoveries from old primary to new primary
-                int numRecoveringReplicas = 0;
                 for (ShardRouting assigned : assignedShards(routing.shardId())) {
                     if (assigned.primary() == false && assigned.initializing() &&
                         assigned.recoverySource().getType() == RecoverySource.Type.PEER) {
-                        numRecoveringReplicas++;
+                        Map<String, Recoveries> recoveriesToUpdate = getReplicaRecoveries(assigned);
+                        Recoveries.getOrAdd(recoveriesToUpdate, routing.relocatingNodeId()).addOutgoing(-1);
+                        Recoveries.getOrAdd(recoveriesToUpdate, routing.currentNodeId()).addOutgoing(1);
                     }
                 }
-                recoveriesPerNode.get(routing.relocatingNodeId()).addOutgoing(-numRecoveringReplicas);
-                recoveriesPerNode.get(routing.currentNodeId()).addOutgoing(numRecoveringReplicas);
+
             }
         }
     }
 
+    private Map<String, Recoveries> getReplicaRecoveries(ShardRouting routing) {
+        if(routing.unassignedReasonIndexCreated() && !routing.primary()) {
+            return initialRecoveriesPerNode;
+        } else {
+            return recoveriesPerNode;
+        }
+    }
+
```

Comment:
The if can return false for primary shard too. Do you want to throw an IllegalArgumentException or add an assertion?